sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -84,26 +91,102 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
   return Program;
 }
 
+template<typename RetT>
+RetT *
+waitUntilBuilt(kernel_program_cache &Cache,
+               kernel_program_cache::EntityWithBuildState<RetT> *WithBuildState) {
+  // condition_variable.wait() call is wrapped with a while due to common
+  // cond_var for building all kernels and programs
+  int state = WithBuildState->State.load();
+  while (state != BS_Built && state != BS_BuildFailed) {
+    Cache.waitUntilBuilt([WithBuildState]() {
+      auto state = WithBuildState->State.load();
+
+      return state == BS_Built || state == BS_BuildFailed;
+    });
+
+    state = WithBuildState->State.load();
+  }
+
+  // expected to be null if state eq BuildFailed
```

Comment:
How the nullptr is handled on the callers side?

Reply:
Now there will be an exception thrown according to spec.